src/library/scala/collection/concurrent/TrieMap.scala

```diff
@@ -1011,6 +1023,40 @@ final class TrieMap[K, V] private (r: AnyRef, rtupd: AtomicReferenceFieldUpdater
   // END extra overrides
   ///////////////////////////////////////////////////////////////////
 
+  override private[collection] def filterInPlaceImpl(p: (K, V) => Boolean): this.type = {
+    val it = readOnlySnapshot().iterator
+    while (it.hasNext) {
+      val (k, v) = it.next()
+      if (p(k, v)) removeRefEq(k, v)
+    }
+    this
+  }
+
+  override private[collection] def mapValuesInPlaceImpl(f: (K, V) => V): this.type = {
+    val it = readOnlySnapshot().iterator
+    while (it.hasNext) {
+      val (k, oldV) = it.next()
+      replaceRefEq(k, oldV, f(k, oldV))
+    }
+    this
+  }
+
+  // overridden to allow using reference equality for checks
+  override def updateWith(key: K)(remappingFunction: Option[V] => Option[V]): Option[V] = rec_updateWith(key)(remappingFunction)
+
+  @tailrec
+  private def rec_updateWith(key: K)(remappingFunction: Option[V] => Option[V]): Option[V] = {
+    val previousValue = this.get(key)
+    val nextValue = remappingFunction(previousValue)
+    (previousValue, nextValue) match {
+      case (None, None) => None
+      case (None, Some(next)) if this.putIfAbsent(key, next).isEmpty => nextValue
+      case (Some(prev), None) if removeRefEq(key, prev) => None
+      case (Some(prev), Some(next)) if replaceRefEq(key, prev, next) => nextValue
+      case _ => this.rec_updateWith(key)(remappingFunction)
```

Comment:
Are you sure this logic is a fast way to do this?  Unless there's pretty clever optimization--possible, but I don't remember!--you'd have to do extra checks on the match.

I'd have thought the nested version would be more likely to be fast, and the `Some` cases would be more likely (in the past, grabbing these first also gives a minor speed benefit because the alternate path doesn't need to check that it's `None`--in theory, optimization could know this and remove the work, but when I was testing things like this I didn't see it in practice).  Something like this:

```scala
val previousValue = get(key)
val nextValue = remappingFunction(previousValue)
previousValue match {
  case Some(prev) => nextValue match {
    case Some(next) => if (replaceRefEq(key, prev, next)) return NextValue
    case _ => if (putIfAbsent(key, next).isEmpty) return nextValue
  }
  case _ => nextValue match {
    case Some(prev) => if (removeRefEq(key, prev)) return None
    case _ => return None
  }
  rec_updateWith(key)(remappingFunction)
}
```

Reply:
Agree this makes sense to write in @Ichoran's style (also in the superclass).

You could even define

```
private[concurrent] def replaceRefEq(k: K, oldValue: V, newValue: V): Boolean = replace(k, oldValue, newValue)
private[concurrent] def removeRefEq(k: K, v: V): Boolean = remove(k, v)
```

and override those in `TrieMap`. Then you wouldn't need to override `updateWith` / `mapValuesInPlaceImpl` / `filterInPlaceImpl`

Reply:
I'm not sure how I feel about a `while (true)` loop—it requires having an unreachable return value, which I think is uglier than the tailrec version.

also, your conditions aren't correct @Ichoran. it looks like they got mixed up when you were rewriting them.

I've otherwise rewritten it in the aforementioned style. I'm happy to change it to use a while loop instead if you want

Reply:
Ack, sorry.  I should be more careful even when writing sketches of solutions (and be clearer that it needs scrutiny).

The while loop vs tail recursion isn't important for speed (that I've ever found, at least not consistently, except in case of nested recursion where while loops optimize better usually starting at 3 deep), just a (mild) style preference on my part, which I recognize is less idiomatic.

Reply:
So--to make the conclusion explicit--yes, leave it as tail recursion!